src/Nethermind/Nethermind.Blockchain/Contracts/Contract.ConstantContract.cs

```diff
@@ -27,87 +28,97 @@ public partial class Contract
             {
                 _contract = contract;
-                _readOnlyTransactionProcessorSource = readOnlyTransactionProcessorSource ?? throw new ArgumentNullException(nameof(readOnlyTransactionProcessorSource));
+                _readOnlyTxProcessorSource = readOnlyTxProcessorSource ?? throw new ArgumentNullException(nameof(readOnlyTxProcessorSource));
             }
-
-            private byte[] Call(BlockHeader parentHeader, string functionName, Transaction transaction)
+            
+            public object[] Call(CallInfo callInfo)
             {
-                lock (_readOnlyTransactionProcessorSource)
+                lock (_readOnlyTxProcessorSource)
                 {
-                    using var readOnlyTransactionProcessor = _readOnlyTransactionProcessorSource.Get(GetState(parentHeader));
-                    return _contract.CallCore(readOnlyTransactionProcessor, parentHeader, functionName, transaction, true);
+                    using var readOnlyTransactionProcessor = _readOnlyTxProcessorSource.Get(GetState(callInfo.ParentHeader));
+                    return CallRaw(callInfo, readOnlyTransactionProcessor);
                 }
             }
 
-            private object[] Call(BlockHeader parentHeader, string functionName, Address sender, params object[] arguments)
-            {
-                return CallRaw(parentHeader, functionName, sender, arguments);
-            }
-
-            /// <summary>
-            /// 
-            /// </summary>
-            /// <param name="parentHeader"></param>
-            /// <param name="functionName"></param>
-            /// <param name="sender"></param>
-            /// <param name="arguments"></param>
-            /// <typeparam name="T"></typeparam>
-            /// <returns></returns>
-            public T Call<T>(BlockHeader parentHeader, string functionName, Address sender, params object[] arguments)
+            protected virtual object[] CallRaw(CallInfo callInfo, IReadOnlyTransactionProcessor readOnlyTransactionProcessor)
             {
-                return (T) Call(parentHeader, functionName, sender, arguments)[0];
+                var transaction = GenerateTransaction(callInfo);
+                if (readOnlyTransactionProcessor.IsContractDeployed(_contract.ContractAddress))
+                {                    
+                    var result = CallCore(callInfo, readOnlyTransactionProcessor, transaction);
+                    return callInfo.Result = _contract.DecodeReturnData(callInfo.FunctionName, result);
+                }
+                else if (callInfo.MissingContractResult != null)
+                {
+                    return callInfo.MissingContractResult;
+                }
+                else
+                {
+                    
+                    throw new AbiException($"Missing contract on address {_contract.ContractAddress} when calling function {callInfo.FunctionName}.");
```

Comment:
generally missing contract should simply behave like return 0x
not sure if this is confusing for any other callers f this API

Reply:
Hm... this wouldn't work well with decoding return values